// rax-pwa/es/Navigation/index.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import { createElement, useEffect, useState, Fragment } from 'rax';
import View from 'rax-view';
import TabBar from '../TabBar/index';
import styles from './index.css';

var _updatePageTrigger = function _updatePageTrigger() {};

var alivePages = [];
var alivePagesCache = {};
var config = {
  maxAlivePageNum: 3,
  pageProps: {},
  routes: []
};
export function activatePageComponent(route) {
  route.component().then(function (fn) {
    return fn();
  }).then(function (Page) {
    if (!route.keepAlive) {
      // ignore page without keepAlive
      return false;
    }

    alivePagesCache[route.path] = createElement(Page, config.pageProps); // Prevent cache from being too large

    if (Object.keys(alivePagesCache).length > config.maxAlivePageNum) {
      delete alivePagesCache[Object.keys(alivePagesCache)[0]];
    }

    _updatePageTrigger(Date.now());
  });
}
;
export function getRoutes() {
  return config.routes;
}
export default function Navigation(props) {
  var appConfig = props.appConfig,
      component = props.component,
      history = props.history,
      routes = props.routes;
  var maxAlivePageNum = appConfig.maxAlivePageNum,
      tabBar = appConfig.tabBar;

  var _useState = useState(null),
      updateTemp = _useState[0],
      setUpdateTemp = _useState[1];

  var Component = component;
  var currentPathname = history.location.pathname;
  var currentPage = routes.find(function (route) {
    return route.path === currentPathname;
  }) || {};
  var isAlivePage = currentPage.keepAlive;
  useEffect(function () {
    history.listen(function () {
      _updatePageTrigger(Date.now());
    }); // Use display control alive page, need get alive page list.

    routes.forEach(function (route) {
      if (route.keepAlive) {
        alivePages.push(route);
      }
    }); // If current page is alive page, need update routes.

    if (isAlivePage) {
      _updatePageTrigger(Date.now());
    }
  }, []); // Props to page component

  var pageProps = {};
  Object.keys(props).forEach(function (key) {
    if (key !== 'appConfig' && key !== 'component') {
      pageProps[key] = props[key];
    }
  });
  config.pageProps = pageProps;
  config.routes = routes;
  _updatePageTrigger = setUpdateTemp;
  maxAlivePageNum && (config.maxAlivePageNum = maxAlivePageNum);
  return createElement(Fragment, null, isAlivePage ? null : createElement(Component, pageProps), alivePages.length > 0 ? createElement(View, {
    style: _extends({}, styles.container, {
      display: isAlivePage ? 'block' : 'none'
    })
  }, alivePages.map(function (alivePage, index) {
    var alivePageRoute = routes.find(function (route) {
      return route.path === alivePage.path;
    });
    var isCurrentPage = alivePageRoute.path === currentPage.path;
    var alivePageComponent = alivePagesCache[alivePageRoute.path] || null;
    if (isCurrentPage && !alivePageComponent) activatePageComponent(alivePageRoute);
    return createElement(View, {
      key: "alivePage" + index,
      style: _extends({}, styles.alivePage, {
        display: isCurrentPage ? 'block' : 'none'
      })
    }, alivePageComponent);
  })) : null, createElement(TabBar, {
    config: tabBar,
    history: history
  }));
}
