// rax-text/lib/index.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _rax = require("rax");

var _universalEnv = require("universal-env");

var _index = _interopRequireDefault(require("./index.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var __rest = void 0 && (void 0).__rest || function (s, e) {
  var t = {};

  for (var p in s) {
    if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0) t[p] = s[p];
  }

  if (s != null && typeof Object.getOwnPropertySymbols === "function") for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
    if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i])) t[p[i]] = s[p[i]];
  }
  return t;
};

var _styleSheet = _index.default;

function _getClassName() {
  var className = [];
  var args = arguments[0];
  var type = Object.prototype.toString.call(args).slice(8, -1).toLowerCase();

  if (type === 'string') {
    args = args.trim();
    args && className.push(args);
  } else if (type === 'array') {
    args.forEach(function (cls) {
      cls = _getClassName(cls).trim();
      cls && className.push(cls);
    });
  } else if (type === 'object') {
    for (var k in args) {
      k = k.trim();

      if (k && args.hasOwnProperty(k) && args[k]) {
        className.push(k);
      }
    }
  }

  return className.join(' ').trim();
}

function _getStyle(classNameExpression) {
  var cache = _styleSheet.__cache || (_styleSheet.__cache = {});

  var className = _getClassName(classNameExpression);

  var classNameArr = className.split(/\s+/);
  var style = cache[className];

  if (!style) {
    style = {};

    if (classNameArr.length === 1) {
      style = _styleSheet[classNameArr[0].trim()];
    } else {
      classNameArr.forEach(function (cls) {
        var value = _styleSheet[cls.trim()];

        if (typeof value === 'object') {
          style = Object.assign(style, _styleSheet[cls.trim()]);
        }
      });
    }

    cache[className] = style;
  }

  return style;
}

var prefixCls = 'rax-text';
var Text = (0, _rax.forwardRef)(function (props, ref) {
  var className = props.className,
      style = props.style,
      numberOfLines = props.numberOfLines,
      children = props.children,
      onPress = props.onPress,
      onClick = props.onClick,
      rest = __rest(props, ["className", "style", "numberOfLines", "children", "onPress", "onClick"]);

  var handleClick = onClick || onPress;
  var lines = typeof numberOfLines === 'string' ? parseInt(numberOfLines, 10) : numberOfLines;
  var textString = '';

  if (children != null) {
    textString = Array.isArray(children) ? children.join('') : children.toString();
  }

  if (_universalEnv.isWeex) {
    return (0, _rax.createElement)("text", _extends({}, rest, {
      ref: ref,
      className: className,
      style: Object.assign({}, _getStyle(className), Object.assign({
        lines: lines
      }, style)),
      onClick: handleClick
    }), textString);
  } else {
    var classNames = [prefixCls, className];

    if (lines) {
      classNames.push(prefixCls + "--overflow-hidden");

      if (lines === 1) {
        classNames.push(prefixCls + "--singleline");
      } else {
        classNames.push(prefixCls + "--multiline");
      }
    }

    return (0, _rax.createElement)("span", _extends({}, rest, {
      ref: ref,
      className: classNames.join(' '),
      style: Object.assign({}, _getStyle(classNames.join(' ')), Object.assign(Object.assign({}, style), {
        webkitLineClamp: lines > 1 ? lines : undefined
      })),
      onClick: handleClick
    }), textString);
  }
});
Text.displayName = 'Text';
var _default = Text;
exports.default = _default;
