// rax-pwa/lib/Document.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _rax = require("rax");

function Document(props) {
  var pageHtml = props.pageHtml,
      pageData = props.pageData,
      title = props.title,
      styles = props.styles,
      scripts = props.scripts;
  return (0, _rax.createElement)("html", null, (0, _rax.createElement)("head", null, (0, _rax.createElement)("meta", {
    charset: "utf-8"
  }), (0, _rax.createElement)("meta", {
    name: "viewport",
    content: "width=device-width, initial-scale=1.0, maximum-scale=1, minimum-scale=1, user-scalable=no"
  }), (0, _rax.createElement)("title", null, title || 'WebApp'), styles.map(function (style) {
    return (0, _rax.createElement)("link", {
      href: style,
      rel: "stylesheet"
    });
  })), (0, _rax.createElement)("body", null, (0, _rax.createElement)("div", {
    id: "root",
    dangerouslySetInnerHTML: {
      __html: pageHtml || ''
    }
  }), scripts.map(function (script) {
    return (0, _rax.createElement)("script", {
      src: script
    });
  }), (0, _rax.createElement)("script", {
    "data-from": "server",
    type: "application/json",
    dangerouslySetInnerHTML: {
      __html: pageData
    }
  })));
}

var _default = Document;
exports.default = _default;
