// rax-pwa/es/Document.js
import { createElement } from 'rax';

function Document(props) {
  var pageHtml = props.pageHtml,
      pageData = props.pageData,
      title = props.title,
      styles = props.styles,
      scripts = props.scripts;
  return createElement("html", null, createElement("head", null, createElement("meta", {
    charset: "utf-8"
  }), createElement("meta", {
    name: "viewport",
    content: "width=device-width, initial-scale=1.0, maximum-scale=1, minimum-scale=1, user-scalable=no"
  }), createElement("title", null, title || 'WebApp'), styles.map(function (style) {
    return createElement("link", {
      href: style,
      rel: "stylesheet"
    });
  })), createElement("body", null, createElement("div", {
    id: "root",
    dangerouslySetInnerHTML: {
      __html: pageHtml || ''
    }
  }), scripts.map(function (script) {
    return createElement("script", {
      src: script
    });
  }), createElement("script", {
    "data-from": "server",
    type: "application/json",
    dangerouslySetInnerHTML: {
      __html: pageData
    }
  })));
}

export default Document;
